[Lens] Range event annotations

mbondyra opened the pull request:
## Summary

Added range event annotations
<img width="976" alt="Screenshot 2022-04-19 at 15 57 13" src="https://user-images.githubusercontent.com/4283304/164020791-9168523e-14bb-46ef-87db-6f84d87e1286.png">

#### Outline annotations


<img width="662" alt="Screenshot 2022-04-21 at 13 12 15" src="https://user-images.githubusercontent.com/4283304/164450300-04a5b715-5542-4365-8509-a745e6c6d1d5.png">
<img width="659" alt="Screenshot 2022-04-21 at 13 12 04" src="https://user-images.githubusercontent.com/4283304/164450308-19dddbbd-4266-475e-a3e9-7379a5cdc3d2.png">

#### Config panel settings
<img width="435" alt="Screenshot 2022-04-19 at 16 16 20" src="https://user-images.githubusercontent.com/4283304/164024729-5ccaccd8-71e6-4957-8412-ff0e1e2b0f26.png">
<img width="487" alt="Screenshot 2022-04-19 at 16 17 07" src="https://user-images.githubusercontent.com/4283304/164024882-d4a21c04-2110-4a3b-af06-a112466145f9.png">

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

MichaelMarcialis reviewed (commented):
This is looking great, @mbondyra! I'm leaving a few comments below for my review. Let me know if you have any questions.

- Would it be possible to reduce the font size of "Apply as range" label text from the current 14px to 12px? This can likely be achieved by wrapping the text in a `<EuiText size="xs">` component.

![image](https://user-images.githubusercontent.com/3884767/163871075-23c42766-e764-4c96-98bc-16a3d91977ed.png)

- It appears that the current fill color for annotation ranges has its opacity hard-coded. I understand we likely did that in order to be consistent with reference lines, as the opacity for reference line fills are hard-coded as well (in an attempt to keep the interface simple and reduce the number of overall inputs required). However, my original intent was that the color selection for ranged annotations would allow users to change the fill opacity, as the circumstances are slightly different here (as color here is only being applied to fill, not both fill and line color as in reference lines). For example, I imagine there may be some cases in which users will wish to use a completely opaque fill color (such as when positioned outside the chart). Any thoughts on allowing the users to edit their opacity via the color picker for annotation ranges?

![image](https://user-images.githubusercontent.com/3884767/163871605-fe2c88e8-60db-417d-b7ec-bb86ba865d36.png)

- The following annotation configuration inputs require the appropriate prop added that allows them to expand to full width: Annotation date from/to, Name, Color, Icon decoration (not technically an annotation range input, but still requires the prop).

![image](https://user-images.githubusercontent.com/3884767/163872487-9f747af5-e07e-46e3-8c4b-13940f73f468.png)

There were also two items that we had discussed as a group via your Google doc, but I'm not sure if we came to a resolution:

- For the initial release, are we unable to show tooltip information for multiple annotation ranges (pending the working being done to redesign the Elastic chart tooltips)? And as such, if an annotation range appears below another annotation range, users will be unable to access that lower layer annotation's tooltip information?

- Did we decide on a desired height for the height of range annotations when positioned outside of the chart? It seems a bit big/tall currently. I had originally assumed it would be short enough that it wouldn't overlap with the tick labels. CCing @gvnmagni in case he has any thoughts.

gvnmagni commented:
> Did we decide on a desired height for the height of range annotations when positioned outside of the chart? It seems a bit big/tall currently. I had originally assumed it would be short enough that it wouldn't overlap with the tick labels. CCing @gvnmagni in case he has any thoughts.

No strong preference here, I would love to have something small and minimal since I'm not a huge fan of overlapping with the time axis but at the same time I'm aware we can't probably avoid it

markov00 commented:
> ❓ opacity for color fill - the color I pass to charts is the exact colour but elastic-charts apply opacity. Pinging marcov here to see if we can do anything about it. 

Wrong Marco here :D 
You can pass the opacity in two ways into the Rect and Line styles prop:
- ` fill: 'rgba(255,255,0,0.8)', opacity: 1` where you use a `rgba`color with its own opacity, but you need to add the `opacity:1` to avoid overriding that opacity
- `fill: 'solid_color' opacity: your_opacity` where you pass a solid color and the opacity as a separate value 

>❓ tooltip information - unfortunately there's nothing we can do. It's the same problem for reference lines. 

:( True we will improve that in the future

> ? outside range annotation height - we decided on the meeting that we'll make it the height of the axis getting this information from the theme. but unfortunately, I cannot find this in the theme. @markov00 maybe I am missing something? 

It depends if you are using the legacy time axis or not. 
- For the legacy time axis the calculation you can just use the tick line size: `theme.tickLine.size`
- For the multilayer time axis, depends on how many rows we are using it (it should be 3 by default in Lens). The formula you can use is `axisStyle.tickLine.padding + theme.tickLabel.fontSize * numberOfLayers`. we could want to cover only the first layer probably so you can reduce it to `axisStyle.tickLine.padding + theme.tickLabel.fontSize`. I've, on purpose, used the `axisStyle` instead of the theme, because the multilayer time axis override the current theme configuration. Please ping me if you need more direction here

mbondyra commented:
@MichaelMarcialis addressed your comments about color opacity and outside range annotations according to @markov00 comments and what we talked about async. Please recheck when you find a moment 🙏🏼

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @mbondyra! I've left two small comments below, but assuming those get addressed, this looks good to me. Approving now so I don't hold you up.

mbondyra commented on src/plugins/chart_expressions/expression_xy/public/components/annotations.tsx:
```diff
@@ -229,6 +259,39 @@ export const Annotations = ({
+                <EuiText size="xs">
+                  <h4>
+                    {formatter
+                      ? `${formatter.convert(time)} — ${formatter?.convert(endTime)}`
+                      : `${moment(time).toISOString()} — ${moment(endTime).toISOString()}`}
```
```suggestion
                      ? `${formatter.convert(time)}–${formatter?.convert(endTime)}`
                      : `${moment(time).toISOString()}–${moment(endTime).toISOString()}`}
```

mbondyra commented on src/plugins/chart_expressions/expression_xy/public/components/annotations.tsx:
```diff
@@ -229,6 +259,39 @@ export const Annotations = ({
+                </EuiText>
+                <EuiText size="xs">{label}</EuiText>
+              </>
+            )}
```
That's because tooltip for rectAnnotations doesn't have a `header` so I was trying to solve it in a different way. Pinging @markov00 again to see if I am not missing something.

markov00 replied:
customTooltip for annotation are freeform actually, but you can just use this CSS for that:
```
 .echTooltip {
  @include euiToolTipStyle;
  @include euiFontSizeXS;
  padding: 0;
  transition: opacity $euiAnimSpeedNormal;
  pointer-events: none;
  user-select: none;
  max-width: 256px;

  &__header {
    @include euiToolTipTitle;
    margin-bottom: 0;
    padding: $euiSizeXS ($euiSizeXS * 2);
  }
```

flash1293 commented on x-pack/plugins/lens/public/xy_visualization/xy_config_panel/color_picker.tsx:
```diff
@@ -47,49 +42,64 @@ export const ColorPicker = ({
+  }, [currentColor]);
 
   const [color, setColor] = useState(currentColor);
 
```
No must have, but I think it's a little weird the default of 10% opacity is dropped in favor of 100% if a color is picked from the swatch. What do you think about keeping the previous alpha value? We can do this by passing in the `swatches` prop and setting it to `euiPaletteColorBlind()` with the opacity of the current color value.

flash1293 commented on x-pack/plugins/lens/public/xy_visualization/xy_config_panel/annotations_config_panel/index.scss:
```diff
@@ -0,0 +1,12 @@
+}
+
+.lnsConfigPanelDate__label {
+  min-width: 56px;
```
After thinking about it a bit I think this is the best solution in this particular scenario. However a value like this looks like an immediate code smell, could we add a comment to explain?

flash1293 reviewed (approved):
Some small suggestions, but overall this LGTM. Great feature!

mbondyra commented:
@elasticmachine merge upstream
